drivers/led/snled27351-spi.c

```diff
@@ -0,0 +1,251 @@
+#endif
+
+#ifndef SNLED27351_CURRENT_TUNE
+#    define SNLED27351_CURRENT_TUNE \
+        { 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF }
+#endif
+
+#define SNLED27351_WRITE (0 << 7)
+#define SNLED27351_READ (1 << 7)
+#define SNLED27351_PATTERN (2 << 4)
+
+#ifdef DRIVER_CS_PINS
+pin_t cs_pins[] = DRIVER_CS_PINS;
+#else
+error "no DRIVER_CS_PINS defined"
+#endif
+
+// These buffers match the snled27351 PWM registers.
+// The control buffers match the PG0 LED On/Off registers.
+// Storing them like this is optimal for I2C transfers to the registers.
+// We could optimize this and take out the unused registers from these
+// buffers and the transfers in snled27351_write_pwm_buffer() but it's
+// probably not worth the extra complexity.
+uint8_t g_pwm_buffer[SNLED27351_DRIVER_COUNT][SNLED27351_PWM_REGISTER_COUNT];
+bool    g_pwm_buffer_update_required[SNLED27351_DRIVER_COUNT] = {false};
+
+uint8_t g_led_control_registers[SNLED27351_DRIVER_COUNT][SNLED27351_LED_CONTROL_REGISTER_COUNT]             = {0};
+bool    g_led_control_registers_update_required[SNLED27351_DRIVER_COUNT] = {false};
+
+
+
+bool snled27351_write(uint8_t index, uint8_t page, uint8_t reg, uint8_t *data, uint8_t len) {
+    static uint8_t spi_transfer_buffer[2] = {0};
+
+    if (index > ARRAY_SIZE(((pin_t[])DRIVER_CS_PINS)) - 1) return false;
+
+    if (!spi_start(cs_pins[index], false, 0, SNLED23751_SPI_DIVISOR)) {
+        spi_stop();
+        return false;
+    }
+
+    spi_transfer_buffer[0] = SNLED27351_WRITE | SNLED27351_PATTERN | (page & 0x0F);
+    spi_transfer_buffer[1] = reg;
+
+    if (spi_transmit(spi_transfer_buffer, 2) != SPI_STATUS_SUCCESS) {
+        spi_stop();
+        return false;
+    }
+
+    if (spi_transmit(data, len) != SPI_STATUS_SUCCESS) {
+        spi_stop();
+        return false;
+    }
+
+    spi_stop();
+    return true;
+}
+
+bool snled27351_write_register(uint8_t index, uint8_t page, uint8_t reg, uint8_t data) {
+    return snled27351_write(index, page, reg, &data, 1);
+}
+
+bool snled27351_write_pwm_buffer(uint8_t index, uint8_t *pwm_buffer) {
+    if (g_pwm_buffer_update_required[index]) {
+        snled27351_write(index, LED_PWM_PAGE, 0, g_pwm_buffer[index], SNLED27351_PWM_REGISTER_COUNT);
+    }
+    g_pwm_buffer_update_required[index] = false;
+    return true;
+}
+
+void snled27351_init_drivers(void) {
+#if defined(LED_DRIVER_SHUTDOWN_PIN)
+    setPinOutput(LED_DRIVER_SHUTDOWN_PIN);
+    writePinHigh(LED_DRIVER_SHUTDOWN_PIN);
+#endif
+
+    spi_init();
+
+    for (uint8_t i = 0; i < SNLED27351_DRIVER_COUNT; i++)
+        snled27351_init(i);
+
+    for (int index = 0; index < SNLED27351_LED_COUNT; index++) {
+        snled27351_set_led_control_register(index, true, true, true);
+    }
+
+    for (uint8_t i = 0; i < SNLED27351_DRIVER_COUNT; i++)
+        snled27351_update_led_control_registers(i);
+}
+
+void snled27351_init(uint8_t index) {
+    setPinOutput(cs_pins[index]);
+    writePinHigh(cs_pins[index]);
```

Comment:
This is the legacy GPIO API and would need to be updated.

Reply:
I see:

`setPinOutput(pin)` -> `gpio_set_pin_output(pin)`
`writePinHigh(pin)` -> `gpio_write_pin_high(pin)`